lib/Runtime/PlatformAgnostic/Platform/Linux/DateTime.cpp

```diff
@@ -13,155 +13,219 @@ namespace PlatformAgnostic
 
-    static inline void SYSTIME_TO_TM(SYSTEMTIME *sysTime, struct tm *time)
-    {
-        time->tm_year = MIN_ZERO(sysTime->wYear - 1900);
-        time->tm_mon = MIN_ZERO(sysTime->wMonth - 1);
-        time->tm_wday = sysTime->wDayOfWeek;
-        time->tm_mday = sysTime->wDay;
-        time->tm_hour = sysTime->wHour;
-        time->tm_min = sysTime->wMinute;
-        time->tm_sec = sysTime->wSecond;
+    const WCHAR *Utility::GetDaylightName(size_t *nameLength, const DateTime::YMD *ymd) {
+        // xplat only gets the actual zone name for the given date
+        return GetStandardName(nameLength, ymd);
     }
 
-    static void SysLocalToUtc(SYSTEMTIME *local, SYSTEMTIME *utc)
+    static void SysLocalToUtc(YMD *local, YMD *utc)
     {
         struct tm local_tm;
-        SYSTIME_TO_TM(local, (&local_tm));
+        bool leap_added;
+        YMD_TO_TM(local, (&local_tm), &leap_added);
 
         // tm doesn't have milliseconds
+        int milliseconds = local->time % 1000;
 
         tzset();
+        int offset;
         time_t utime = timegm(&local_tm);
-        mktime(&local_tm); // we just want the gmt_off, don't care the result
-        utime -= local_tm.tm_gmtoff; // reverse UTC
+
+        // mktime min year 1900
+        if (local_tm.tm_year < 1900)
+        {
+            struct tm tm = local_tm;
+            tm.tm_year = 1900;
+            mktime(&tm);
+            offset = tm.tm_gmtoff;
+        }
+        else
+        {
+            mktime(&local_tm);
+            offset = local_tm.tm_gmtoff;
+        }
+        utime -= offset;
 
         struct tm utc_tm;
         if (gmtime_r(&utime, &utc_tm) == 0)
         {
             AssertMsg(false, "gmtime() failed");
         }
 
-        TM_TO_SYSTIME((&utc_tm), utc);
+        TM_TO_YMD((&utc_tm), utc, leap_added, local->year);
         // put milliseconds back
-        utc->wMilliseconds = local->wMilliseconds;
+        utc->time += milliseconds;
     }
 
-    static void SysUtcToLocal(SYSTEMTIME *utc, SYSTEMTIME *local,
+    static void SysUtcToLocal(YMD *utc, YMD *local,
                           int &bias, int &offset, bool &isDaylightSavings)
     {
         struct tm utc_tm;
-        SYSTIME_TO_TM(utc, (&utc_tm));
+        bool leap_added;
+        YMD_TO_TM(utc, &utc_tm, &leap_added);
 
         // tm doesn't have milliseconds
+        int milliseconds = utc->time % 1000;
 
         tzset();
         time_t ltime = timegm(&utc_tm);
         struct tm local_tm;
         localtime_r(&ltime, &local_tm);
-        offset = local_tm.tm_gmtoff / 60;
-
-        TM_TO_SYSTIME((&local_tm), local);
 
+        TM_TO_YMD((&local_tm), local, leap_added, utc->year);
         // put milliseconds back
-        local->wMilliseconds = utc->wMilliseconds;
-
-        isDaylightSavings = local_tm.tm_isdst;
+        local->time += milliseconds;
+
+        // ugly hack but;
+        // right in between dst pass
+        // we need mktime trick to get the correct dst
+        utc_tm.tm_isdst = 1;
```

Comment:
Instead of doing this, can you just use the timezone variable that is set by tzset?